mmlearn/tasks/linear_probing.py

```diff
@@ -0,0 +1,454 @@
+                Dict[str, partial[torch.optim.lr_scheduler.LRScheduler]],
+                partial[torch.optim.lr_scheduler.LRScheduler],
+            ]
+        ] = None,
+    ):
+        super().__init__()
+        assert task in ["binary", "multiclass", "multilabel"], (
+            f"Invalid task type: {task}. "
+            "Expected one of ['binary', 'multiclass', 'multilabel']."
+        )
+
+        self.modality = modality
+
+        self.encoder: nn.Module = extract_vision_encoder(
+            encoder, model_checkpoint_path, keys_to_rename={"encoders.rgb.": ""}
+        )
+
+        linear_layer = MLP(num_output_features, num_classes, hidden_dims)
+
+        if pre_classifier_batch_norm:
+            linear_layer = nn.Sequential(
+                nn.BatchNorm1d(num_output_features, affine=False),
+                linear_layer,
+            )
+        self.classifier = linear_layer
+
+        self.freeze_encoder = freeze_encoder
+        self.num_classes = num_classes
+
+        if self.freeze_encoder:
+            for param in self.encoder.parameters():
+                param.requires_grad = False
+
+        self.loss_fn = nn.CrossEntropyLoss()
+
+        self.top_k_list = top_k_list
+        if task == "multiclass":
+            if self.top_k_list is None:
+                self.top_k_list = [1, 5]
+            accuracy_metrics = {
+                f"top_{k}_accuracy": Accuracy(
+                    task=task, num_classes=num_classes, top_k=k
+                )
+                for k in self.top_k_list
+            }
+
+            # Additional metrics for multiclass classification
+            additional_metrics = {
+                "precision": Precision(
+                    task=task, num_classes=num_classes, average="macro"
+                ),
+                "recall": Recall(task=task, num_classes=num_classes, average="macro"),
+                "f1_score": F1Score(
+                    task=task, num_classes=num_classes, average="macro"
+                ),
+                "auc": AUROC(
+                    task=task, num_classes=num_classes, average="macro"
+                ),  # AUROC for multiclass
+            }
+
+        elif task == "multilabel":
+            # Accuracy and other metrics for multilabel classification
+            accuracy_metrics = {"accuracy": Accuracy(task=task, num_labels=num_classes)}
+
+            # Additional metrics for multilabel classification
+            additional_metrics = {
+                "precision": Precision(
+                    task=task, num_labels=num_classes, average="macro"
+                ),
+                "recall": Recall(task=task, num_labels=num_classes, average="macro"),
+                "f1_score": F1Score(task=task, num_labels=num_classes, average="macro"),
+                "auc": AUROC(task=task, num_labels=num_classes),  # AUC for multilabel
+            }
+
+        else:  # binary
+            # Accuracy and other metrics for binary classification
```

Comment:
For creating metrics, try to make [this](https://github.com/VectorInstitute/mmlearn/blob/main/mmlearn/tasks/zero_shot_classification.py#L248) a standalone function that can be reused by the `LinearClassifier` and `ZeroShotClassification` classes. Currently, you can directly reuse it by calling `ZeroShotClassification._create_metrics...`.